### PR TITLE
Add workflow builders

### PR DESCRIPTION
**Pull request by nolag**
Last part taken from [this PR](https://github.com/smartcontractkit/chainlink-common/pull/695/files)

### CODE REVIEW

File: .golangci.yml
Comment (jmank88) [inline .golangci.yml:1]: Is this third party exclusion still necessary after the other changes?

File: go.mod [hunk 1]
```diff
@@ -42,14 +47,23 @@ require (
...
+require (
+	github.com/goccy/go-yaml v1.11.3 // indirect
+	github.com/inconshreveable/mousetrap v1.1.0 // indirect
+	github.com/mitchellh/go-wordwrap v1.0.1 // indirect
+	github.com/sanity-io/litter v1.5.5 // indirect
+	github.com/spf13/pflag v1.0.5 // indirect
+	golang.org/x/xerrors v0.0.0-20231012003039-104605ab7028 // indirect
+)
+
 require (
```
Comment (jmank88) [inline go.mod:59]:
```suggestion
```

File: pkg/workflows/builder.go [hunk 2]
```diff
@@ -1,35 +1,184 @@
...
-type WorkflowSpecFactory struct{}
+	"github.com/smartcontractkit/chainlink-common/pkg/capabilities"
+)
+
+// 1. Cap defines JSON schema for inputs and outputs of a cap.
+// Trigger: triggerOutputType := workflowBuilder.addTrigger(DataStreamsTrigger.ModifiedConfig{})
+// Adds metadata to the builder. Returns output type.
+// 2. Consensus: consensusOutputType := workflowBuilder.addConsensus(ConsensusConfig{
+// 	Inputs: triggerOutputType,
+// })
```
Comment (jmank88) [inline pkg/workflows/builder.go:16]: Detached comments don't render will in docs. Should this be a package level comment at the top of the file?

File: pkg/workflows/builder.go [hunk 3]
```diff
@@ -1,35 +1,184 @@
...
+
+type WorkflowSpecFactory struct {
+	spec           *WorkflowSpec
```
Comment (jmank88) [inline pkg/workflows/builder.go:19]: How about just `SpecFactory` and even `Spec` since we are already in `package workflows` and callers do `workflows.WorkflowSpecFactory` & `workflows.WorkflowSpec`.
```suggestion
type SpecFactory struct {
	spec           *Spec
```

**Review by jmank88 — APPROVED** [review]

File: pkg/workflows/builder.go (hunk 3 above)
Comment (nolag) [reply pkg/workflows/builder.go:19]: Let's do that one in another PR since WorkflowSpec existed before any of the builder work.  In a follow up I can rename both together.

File: pkg/workflows/builder.go (hunk 2 above)
Comment (nolag) [reply pkg/workflows/builder.go:16]: It is an old comment from when @DeividasK and I were exploring that never got removed, so I'll remove it.

**Review by jmank88 — DISMISSED** [review]

File: pkg/workflows/builder.go [hunk 4]
```diff
@@ -1,35 +1,177 @@
...
+	return &capDefinitionImpl[O]{ref: originalRef[:len(originalRef)-1] + "." + fieldName + ")"}
+}
+
+func (w *WorkflowSpecFactory) Spec() (WorkflowSpec, error) {
+	if len(w.duplicateNames) > 0 {
+		duplicates := make([]string, 0, len(w.duplicateNames))
+		for k := range w.duplicateNames {
+			duplicates = append(duplicates, k)
+		}
+		return WorkflowSpec{}, fmt.Errorf("duplicte step ids %v", strings.Join(duplicates, ", "))
```
Comment (bolekk) [inline pkg/workflows/builder.go:156]: nit: typo

File: pkg/workflows/builder_test.go [hunk 5]
```diff
@@ -0,0 +1,398 @@
...
+	"github.com/smartcontractkit/chainlink-common/pkg/capabilities/consensus/ocr3"
+	"github.com/smartcontractkit/chainlink-common/pkg/capabilities/targets/chainwriter"
+	"github.com/smartcontractkit/chainlink-common/pkg/capabilities/triggers/streams"
+	"github.com/smartcontractkit/chainlink-common/pkg/workflows"
+	"github.com/smartcontractkit/chainlink-common/pkg/workflows/testdata/fixtures/capabilities/notstreams"
+)
+
+//go:generate go run github.com/smartcontractkit/chainlink-common/pkg/capabilities/cli/cmd/generate-types --dir $GOFILE
+
+// Note that the set of tests in this file cover the conversion from existing YAM -> code
```
Comment (bolekk) [inline pkg/workflows/builder_test.go:22]: nit: typo "YAM"

File: pkg/workflows/builder_test.go [hunk 6]
```diff
@@ -0,0 +1,398 @@
...
+	consensus := conf.Ocr.New(workflow, "ccip_feeds", ocr3.ConsensusInput{
+		Observations: workflows.ListOf[[]streams.Feed](streamsTrigger)},
+	)
+
+	conf.ChainWriter.New(workflow, conf.TargetChain, chainwriter.TargetInput{SignedReport: consensus})
+
+	return workflow, nil
+}
+
+// What if there were hundreds of feeds?  Like feeds that aren't for CCIP?
```
Comment (bolekk) [inline pkg/workflows/builder_test.go:51]: Is this comment relevant to any particular field here? Maybe remove?

**Review by bolekk — APPROVED** [review]

File: pkg/workflows/builder_test.go (hunk 6 above)
Comment (nolag) [reply pkg/workflows/builder_test.go:51]: It was for my demo to describe what one of the benefits of code is.  I thought leaving it shows how it can shrink the YAML and be less error-prone.  I'll remove it.

**Review by jmank88 — APPROVED** [review]

**Review by DeividasK — APPROVED** [review]

**Review by bolekk — APPROVED** [review]
